benchmarks/Sources/blackHole.swift

```diff
@@ -0,0 +1,31 @@
+// ===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Collections open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+// ===----------------------------------------------------------------------===//
+
+/// A function to foil compiler optimizations that would otherwise optimize out code you want to benchmark.
+///
+/// The function wraps another object or function, does nothing, and returns.
+/// If you want to benchmark the time is takes to create an instance and you don't maintain a reference to it, the compiler may optimize it out entirely, thinking it is unused.
+/// To prevent the compiler from removing the code you want to measure, wrap the creation of the instance with `blackHole`.
+/// For example, the following code benchmarks the time it takes to create an instance of `Date`, and wraps the creation of the instance to prevent the compiler from optimizing it away:
+///
+/// ```swift
+/// Benchmark("Foundation Date()",
+///     configuration: .init(
+///         metrics: [.throughput, .wallClock],
+///         scalingFactor: .mega)
+/// ) { benchmark in
+///     for _ in benchmark.scaledIterations {
+///         blackHole(Date())
+///     }
+/// }
+/// ```
+@inline(never)
+public func blackHole(_: some Any) {}
```

Comment:
This blackHole might be a bit problematic? It'd be compiled together in the same module as the code under measurement, so it might not work as expected, or am I missing something? Shouldn't it go to a separate target/module ?

Reply:
One thing to look for is if you get a runtime that is extremely short, on my machine that benchmark executes on a wall clock time of ~41ns - exactly the same as I'll get if I simply comment out the call to tinyArrayNonAllocationFunctions() - even if I put in an inner loop there, it still takes ~41ns (which seems to be the overhead of a function call?).

I'd probably suggest a small restructuring something like this:
```swift
import _CertificateInternals

public func tinyArrayNonAllocationFunctions(_ iterations: Range<Int>) -> Int {
    var counts = 0
    for _ in iterations {
        counts += _TinyArray(CollectionOfOne(1)).count

        var array = _TinyArray<Int>()
        array.append(contentsOf: CollectionOfOne(1))
        counts += array.count
    }
    return counts
}

with a call site from the benchmark code:

    Benchmark("TinyArray non-allocating functions") { benchmark in
        blackHole(tinyArrayNonAllocationFunctions(benchmark.scaledIterations)) // this uses the benchmarks blackHole
    }
```

Reply:
See also https://github.com/ordo-one/package-benchmark/issues/178